pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -1215,6 +1217,82 @@ func (p *Pod) ReclaimablePods() ([]kueue.ReclaimablePod, error) {
 	return result, nil
 }
 
+func (p *Pod) CustomWorkloadConditions(wl *kueue.Workload) ([]metav1.Condition, bool) {
+	var (
+		conditions           []metav1.Condition
+		needToUpdateWorkload bool
+	)
+	if condition, updated := p.waitingForReplacementPodsCondition(wl); condition != nil {
+		conditions = append(conditions, *condition)
+		if updated {
+			needToUpdateWorkload = true
+		}
+	}
+	return conditions, needToUpdateWorkload
+}
+
+func (p *Pod) waitingForReplacementPodsCondition(wl *kueue.Workload) (*metav1.Condition, bool) {
+	replCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadWaitingForReplacementPods)
+	replCondStatus := p.hasNotRunnableNorSucceededPods()
+
+	// Nothing to change.
+	if replCond == nil && !replCondStatus {
+		return nil, false
+	}
+
+	var updated bool
+
+	if replCond == nil {
+		replCond = &metav1.Condition{
+			Type: kueue.WorkloadWaitingForReplacementPods,
+		}
+		updated = true
+	}
+
+	if replCondStatus {
+		if evictCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadEvicted); evictCond != nil && evictCond.Status == metav1.ConditionTrue {
+			if replCond.Reason != evictCond.Reason {
+				replCond.Reason = evictCond.Reason
+				updated = true
+			}
+			if replCond.Message != evictCond.Message {
+				replCond.Message = evictCond.Message
+				updated = true
+			}
+		} else if replCond.Status != metav1.ConditionTrue {
+			replCond.Reason = kueue.WorkloadPodsReady
+			replCond.Message = "Has not runnable or not succeeded pods"
+			updated = true
+		}
+
+		if replCond.Status != metav1.ConditionTrue {
+			replCond.Status = metav1.ConditionTrue
+			updated = true
+		}
+	} else if replCond.Status != metav1.ConditionFalse {
+		replCond.Status = metav1.ConditionFalse
+		updated = true
+	}
+
+	if updated {
+		replCond.ObservedGeneration = wl.Generation
+		replCond.LastTransitionTime = metav1.Now()
+	}
+
+	return replCond, updated
+}
+
+// hasNotRunnableNorSucceededPods returns true if it has at least one not runnable
```

Comment:
I prefer we base this on counters.

If somehow a Failed pod disappears, and all other pods are Succeeded, we are still waiting for a replacement Pod.

Reply:
Fixed